source/opt/liveness.cpp

```diff
@@ -0,0 +1,328 @@
+// Copyright (c) 2022 The Khronos Group Inc.
+// Copyright (c) 2022 LunarG Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/opt/liveness.h"
+
+#include "source/opt/ir_context.h"
+
+namespace {
+
+const uint32_t kDecorationLocationInIdx = 2;
+const uint32_t kOpDecorateMemberMemberInIdx = 1;
+const uint32_t kOpDecorateMemberLocationInIdx = 3;
+const uint32_t kOpDecorateBuiltInLiteralInIdx = 2;
+const uint32_t kOpDecorateMemberBuiltInLiteralInIdx = 3;
+
+}  // namespace
+
+namespace spvtools {
+namespace opt {
+namespace analysis {
+
+LivenessManager::LivenessManager(IRContext* ctx) : ctx_(ctx), computed_(false) {
+  // Liveness sets computed when queried
+}
+
+void LivenessManager::InitializeAnalysis() {
+  live_locs_.clear();
+  live_builtins_.clear();
+  // Mark all builtins live for frag shader.
+  if (context()->GetStage() == SpvExecutionModelFragment) {
```

Comment:
Sorry. I probably didn't explain this well enough in the comments. I will improve the comment for IsAnalyzedBuiltin.

Essentially, by my reckoning, there are only three builtins that can be analyzed and removed between two stages: PointSize, ClipDistance and CullDistance. And only if Frag is not the downstream stage. That's because Frag fixed function consumes these values. Any other builtins cannot be removed because they are consumed by the downstream stage's fixed function.

FragDepth is not marked live only because we don't consider removing it in the first place. ClipDistance is considered live in a Frag shader by default because it is consumed by the Frag shader fixed function.

Reply:
Oh ok, thanks for the explanation! Would you have a reference for that?

Reply:
thanks for the time taken to explain!